### PR TITLE
Methods to deal with overplotting large datasets: scattermore and binning

### PR DESCRIPTION
**Pull request by lambdamoses**
I added support for scattermore and binning to `plotColData`, `plotExpression`, and `plotReducedDim`. I also added new unit tests, whose code you can run to see what the plots look like. I ran `R CMD check` and the changes didn't break existing code. I'm not sure how helpful summarizing say gene expression values in bins in `plotColData` and PCA plots would be, though I find it useful in spatial transcriptomics when plotting in histological space, so up to you if you want to keep that part. Addressing #191

### CODE REVIEW

File: R/plotColData.R [hunk 1]
```diff
@@ -2,35 +2,71 @@
...
 #' @param by_exprs_values Alias for \code{by_assay_name}.
-#' @param other_fields Additional cell-based fields to include in the data.frame, see \code{?"\link{scater-plot-args}"} for details.
+#' @param other_fields Additional cell-based fields to include in the
+#'   data.frame, see \code{?"\link{scater-plot-args}"} for details.
 #' @param swap_rownames Column name of \code{rowData(object)} to be used to
-#'  identify features instead of \code{rownames(object)} when labelling plot
-#'  elements.
+#'   identify features instead of \code{rownames(object)} when labelling plot
+#'   elements.
 #' @param color_by Alias to \code{colour_by}.
```
Comment (alanocallaghan) [inline R/plotColData.R]: Can you remove the unrelated doc changes? Thanks

**Review by alanocallaghan — CHANGES_REQUESTED** [review]
Code changes look good, can you either remove the unrelated doc changes or resolve the merge conflicts?

File: R/plotColData.R (hunk 1 above)
Comment (lambdamoses) [reply R/plotColData.R]: I did and resolved the conflicts

**Comment by alanocallaghan** [comment]
Thanks! Can you provide some examples of the binning functionality and scattermore as well? Can drop them here or in the roxygen docs, whichever's easier for you.

**Comment by alanocallaghan** [comment]
Ah, I missed the comment about examples in unit tests. Sorry about that

**Comment by alanocallaghan** [comment]
Going to render this reprex later when test pass

```r
reprex::reprex({
devtools::load_all("~/Documents/github/R/scater")

example_sce <- mockSCE(ncells = 1e5)
example_sce <- logNormCounts(example_sce)
example_sce <- runPCA(example_sce)


plotColData(example_sce, x = "sum", y = "detected", scattermore = FALSE, point_size = 2)
plotColData(example_sce, x = "sum", y = "detected", scattermore = TRUE, point_size = 2)
plotColData(example_sce, x = "sum", y = "detected", scattermore = TRUE, point_size = 2, colour_by = "sizeFactor")
plotColData(example_sce, x = "sum", y = "detected", scattermore = TRUE, shape_by = "Treatment")

plotColData(example_sce, x = "sum", y = "detected", bins = 10)
plotColData(example_sce, x = "sum", y = "detected", bins = 10, hex = TRUE)
plotColData(example_sce, x = "sum", y = "detected", bins = 10, colour_by = "total")
plotColData(example_sce, x = "sum", y = "detected", bins = 10, colour_by = "total", hex = TRUE)
plotColData(example_sce, x = "sum", y = "detected", bins = 10, colour_by = "Treatment")


plotExpression(example_sce, "Gene_0001", x = "Gene_0100", scattermore = TRUE, point_size = 2)
plotExpression(example_sce, "Gene_0001", x = "Gene_0100", scattermore = TRUE, point_size = 2, color_by = "Cell_Cycle")

plotExpression(example_sce, "Gene_0001", x = "Gene_0100", bins = 10)
plotExpression(example_sce, "Gene_0001", x = "Gene_0100", bins = 10, hex = TRUE)
plotExpression(example_sce, "Gene_0001", x = "Gene_0100", bins = 10, colour_by = "Gene_0002", summary_fun = "mean")
plotExpression(example_sce, "Gene_0001", x = "Gene_0100", bins = 10, colour_by = "Gene_0002", hex = TRUE, summary_fun = "mean")

plotPCA(example_sce, ncomponents = 4, scattermore = TRUE, point_size = 3)
plotPCA(example_sce, ncomponents = 4, scattermore = TRUE, point_size = 3, colour_by = "Gene_0001")

plotPCA(example_sce, ncomponents = 4, bins = 10)
plotPCA(example_sce, ncomponents = 4, bins = 10, hex = TRUE)
plotPCA(example_sce, ncomponents = 4, bins = 10, colour_by = "Gene_0001")
plotPCA(example_sce, ncomponents = 4, bins = 10, colour_by = "Gene_0001", hex = TRUE)

plotPCA(example_sce, scattermore = TRUE, point_size = 2)
plotPCA(example_sce, scattermore = TRUE, point_size = 2, colour_by = "Cell_Cycle")

# Binning
plotPCA(example_sce, bins = 10)
plotPCA(example_sce, bins = 10, hex = TRUE)
plotPCA(example_sce, bins = 10, colour_by = "Gene_0001")
plotPCA(example_sce, bins = 10, colour_by = "Gene_0001", hex = TRUE)


})
```

**Comment by lambdamoses** [comment]
I'll add the examples. I'll also change `assay_name` back to `assay.type` to be consistent with other functions in the package.

**Comment by lambdamoses** [comment]
I added the examples and ran R CMD check. On my computer it didn't get errors or warnings. There were 3 notes unrelated to my changes. 
```
❯ checking package dependencies ... NOTE
  Imports includes 21 non-default packages.
  Importing from so many packages makes the package vulnerable to any of
  them becoming unavailable.  Move as many as possible to Suggests and
  use conditionally.

❯ checking installed package size ... NOTE
    installed size is  6.5Mb
    sub-directories of 1Mb or more:
      doc   5.9Mb

❯ checking R code for possible problems ... [30s/30s] NOTE
  .get_point_args: no visible global function definition for ‘modifyList’
  paired_reddim_plot: no visible binding for global variable ‘..scaled..’
  Undefined global functions or variables:
    ..scaled.. modifyList
  Consider adding
    importFrom("utils", "modifyList")
  to your NAMESPACE file.
```

**Comment by alanocallaghan** [comment]
Thanks, the first two notes are fine, the third is my fault, but would be good to add it to the roxygen imports. Will have a look at the outputs later to see if there's anything worth tweaking.

**Comment by alanocallaghan** [comment]
Sorry yeah have been off/away the last couple of weeks

**Comment by alanocallaghan** [comment]
Thanks!

**Comment by lambdamoses** [comment]
You merged it into `master`. Is it going into Bioconductor 3.17? I'm asking because I need to decide if I need to deprecate `plotColDataBin2D()` in Voyager.

**Comment by alanocallaghan** [comment]
I then merged master into devel and pushed to bioc, it should show up on the next build report

**Comment by alanocallaghan** [comment]
Seems to be on bioc now
